File: webapp/Locus.Core/Services/RequirementService.cs
using Locus.Core.DTO;
using Locus.Core.Models;
using Locus.Core.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

//Service for Requirement operations.
namespace Locus.Core.Services
{
    //Interface which contains methods for each CRUD operation
    public interface IRequirementService
    {
        Requirement Save(Requirement requirement, string user);

        Requirement Update(Requirement requirement, string user);

        Requirement Delete(int idRequirement, string user);

        List<Requirement> GetAll();

        Requirement Get(int idRequirement);

        List<TestCase> GetAllTestCase(int idReq);

        List<TestCase> GetAllTestCaseInactives(int idReq);

        List<Requirement> GetProject(int Project_Id);

        List<Requirement> GetProjectInactives(int Project_Id);

        List<TestScenario> GetAllTestScenario(int id);

        List<TestScenario> GetAllTestScenarioInactives(int id);

        List<TestProcedure> GetAllTestProcedure(int id);

        List<TestProcedure> GetAllTestProcedureInactives(int id);

        List<TestProcedure> GetAllTestProcedureNoTc(int id);

        Requirement Enable(int id, string user);

        List<ChangeLog> RequirementChangeLogs(int id);

        ChangeLog Restore(Requirement requirement, int version, string user);

        DashboardDTO GetDashboard(int projectId);

        Requirement GetRequirementbyTCId(int tcId);

        Requirement GetRequirementbyTPId(int tpId);

        List<TestCase> GetAllTCWithoutTP(int projectId);

        List<TestCase> GetAllTCWithoutTPByReq(int reqId);

        List<EvidenceDTO> GetEvidenceFromReq(int[] requirements);

        List<TestDTO> GetTestCases(int projectId);

        List<TestProcedure> GetAllTestProcedureByArray(RequirementDTO[] data);

        List<Requirement> GetRequirementsByTestEvidence(int id, string type);

        List<TestDTO> GetTestProcedures(int projectId);

        List<TestDTO> GetTestScenarios(int projectId);

        Requirement GetByReqNumber(int projectId, string reqnumber);

        List<string> GetSprints(int projectId);

        List<Requirement> GetNonActivesExecutedForTP(int executionId);
        List<Requirement> GetNonActivesExecutedForTC(int executionId);
        List<Requirement> GetNonActivesExecutedForTS(int executionId);

        List<TestCase> GetUniqueTestCases(int ProjectId);
        List<EvidenceDTO> GetAutomatedEvidenceFromReq(int[] ids);
        List<TestDTO> GetTestCasesByNumber(string[] TcNumbers, int projectId, int scriptId);
    }
    //Class which implements IRequirementService's methods and instance to IRequirementRepository
    public class RequirementService : IRequirementService
    {
        //Instance of IRequirementRepository
        private readonly IRequirementRepository _requirementRepository;
        private readonly ITestSuplementalRepository _testSuplementalRepository;
        

        //Constructor of RequirementService and initialize requirementRepository
        public RequirementService(IRequirementRepository requirementRepository)
        {
            _requirementRepository = requirementRepository;
        }

      
        public RequirementService(IRequirementRepository requirementRepository, ITestSuplementalRepository procedureSuplementalRepository)
        {
            _requirementRepository = requirementRepository;
            _testSuplementalRepository = procedureSuplementalRepository;
        }

        //Delete method calls to Delete Method From requirementRepository.
        public Requirement Delete(int idRequirement, string user)
        {
            return _requirementRepository.Delete(idRequirement, user);
        }

        public Requirement Enable(int id, string user)
        {
            return _requirementRepository.Enable(id, user);
        }

        //Get method calls to Get Method From requirementRepository.
        public Requirement Get(int idRequirement)
        {
            return _requirementRepository.Get(idRequirement);
        }
        //GetAll method calls to GetAll Method From requirementRepository.
        public List<Requirement> GetAll()
        {
            return _requirementRepository.GetAll();
        }

        public List<TestCase> GetAllTestCase(int idReq)
        {
            return _requirementRepository.GetAllTestCase(idReq);
        }

        public List<TestCase> GetAllTestCaseInactives(int idReq)
        {
            return _requirementRepository.GetAllTestCaseInactives(idReq);
        }

        public List<TestProcedure> GetAllTestProcedure(int id)
        {
            return _requirementRepository.GetAllTestProcedure(id);
        }

        public List<TestProcedure> GetAllTestProcedureInactives(int id)
        {
            return _requirementRepository.GetAllTestProcedureInactives(id);
        }

        public List<TestProcedure> GetAllTestProcedureNoTc(int id)
        {
            return _requirementRepository.GetAllTestProcedureNoTc(id);
        }

        public List<TestScenario> GetAllTestScenario(int id)
        {
            return _requirementRepository.GetAllTestScenario(id);
        }

        public List<TestScenario> GetAllTestScenarioInactives(int id)
        {
            return _requirementRepository.GetAllTestScenarioInactives(id);
        }

        public DashboardDTO GetDashboard(int projectId)
        {
            DashboardDTO dashboard = new DashboardDTO();
            List<Requirement> reqList = _requirementRepository.GetProject(projectId);
            List<TestSuplemental> suplementals = _testSuplementalRepository.GetForProject(projectId);
            dashboard.totalRequirements = reqList.Count;
            int totaltc = 0;
            int totaltp = 0;
            int totalts = 0;
            int missingtc = 0;
            List<Requirement> reqListMissingEvidence = _requirementRepository.GetAllReqWithoutEvidence(projectId);
            List<TestCase> testCases = _requirementRepository.GetUniqueTestCases(projectId);
            totaltc = testCases.Count;

            List<TestProcedure> testProcedures = _requirementRepository.GetUniqueTestProcedures(projectId);
            totaltp = testProcedures.Count;

            List<TestScenario> testScenarios = _requirementRepository.GetUniqueTestScenarios(projectId);
            totalts = testScenarios.Count;

            missingtc = reqListMissingEvidence.Count;

            dashboard.requirementsMissingTestEvidence = missingtc;
            dashboard.totalstp = _testSuplementalRepository.GetForProject(projectId).Count; 
            dashboard.totalTestCases = totaltc;
            dashboard.totalTestProcedures = totaltp;
            dashboard.totalTestScenarios = totalts;

            return dashboard;


        }

        public List<Requirement> GetProject(int Project_Id)
        {
            return _requirementRepository.GetProject(Project_Id);
        }

        public List<Requirement> GetProjectInactives(int Project_Id)
        {
            return _requirementRepository.GetProjectInactives(Project_Id);
        }

        public List<ChangeLog> RequirementChangeLogs(int id)
        {
            return _requirementRepository.RequirementChangeLogs(id);
        }

        public ChangeLog Restore(Requirement requirement, int version, string user)
        {
            return _requirementRepository.Restore(requirement, version, user);
        }

        //Save method calls to Save Method From requirementRepository.
        public Requirement Save(Requirement requirement, string user)
        {
            return _requirementRepository.Save(requirement, user);
        }
        //Update method calls to Update Method From requirementRepository.
        public Requirement Update(Requirement requirement, string user)
        {
            return _requirementRepository.Update(requirement, user);
        }

        public Requirement GetRequirementbyTCId(int tcId)
        {
            var requirement = _requirementRepository.GetRequirementbyTCId(tcId);
            return requirement;
        }

        public Requirement GetRequirementbyTPId(int tpId)
        {
            var requirement = _requirementRepository.GetRequirementbyTPId(tpId);
            return requirement;
        }

        public List<TestCase> GetAllTCWithoutTP(int projectId)
        {
            return _requirementRepository.GetAllTCWithoutTP(projectId);
        }

        public List<TestCase> GetAllTCWithoutTPByReq(int reqId)
        {
            return _requirementRepository.GetAllTCWithoutTPByReq(reqId);
        }

        public List<TestProcedure> GetAllTestProcedureByArray(RequirementDTO[] data)
        {
            return _requirementRepository.GetAllTestProcedureByArray(data);
        }

        public List<EvidenceDTO> GetEvidenceFromReq(int[] requirements)
        {
            return _requirementRepository.GetEvidenceFromReq(requirements);

        }

        public List<Requirement> GetRequirementsByTestEvidence(int id, string type)
        {
            return _requirementRepository.GetRequirementsByTestEvidence(id, type);
        }

        public List<TestDTO> GetTestCases(int projectId)
        {
            return _requirementRepository.GetTestCases(projectId);
        }

        public List<TestDTO> GetTestProcedures(int projectId)
        {
            return _requirementRepository.GetTestProcedures(projectId);
        }

        public List<TestDTO> GetTestScenarios(int projectId)
        {
            return _requirementRepository.GetTestScenarios(projectId);
        }

        public Requirement GetByReqNumber(int projectId, string reqnumber)
        {
            return _requirementRepository.GetByReqNumber(projectId, reqnumber);
        }
        public List<EvidenceDTO> GetAutomatedEvidenceFromReq(int[] ids)
        {
            return _requirementRepository.GetAutomatedEvidenceFromReq(ids);
        }

        public List<string> GetSprints(int projectId)
        {
            return _requirementRepository.GetSprints(projectId);
        }

        public List<Requirement> GetNonActivesExecutedForTP(int executionId)
        {
            return _requirementRepository.GetNonActivesExecutedForTP(executionId);
        }


        public List<TestCase> GetUniqueTestCases(int ProjectId)
        {
            return _requirementRepository.GetUniqueTestCases(ProjectId);
        }

        public List<Requirement> GetNonActivesExecutedForTC(int executionId)
        {
            return _requirementRepository.GetNonActivesExecutedForTC(executionId);
        }

        public List<Requirement> GetNonActivesExecutedForTS(int executionId)
        {
            return _requirementRepository.GetNonActivesExecutedForTS(executionId);
        }

        public List<TestDTO> GetTestCasesByNumber(string[] TcNumbers, int projectId, int scriptId)
        {
            return _requirementRepository.GetTestCasesByNumber(TcNumbers, projectId, scriptId);
        }
    }
}

File: webapp/Locust/Scripts/app/services/RequirementService.js
(function (angular, undefined) {
    'use strict';

    // Get services module
    var app = angular.module(appName);

    app.service('RequirementService', ['$resource', function ($resource) {

        var noop = angular.noop;
        var resource = $resource('/api/Requirement/:action/:id',
            { id: '@Id' }, {
                Save: {
                    method: 'POST'
                },
                Update: {
                    method: 'PUT'
                },
                Delete: {
                    method: 'DELETE',
                    params: {
                        id: "@id"
                    }
                },
                Enable: {
                    method: 'GET',
                    isArray: false
                },
                Get: {
                    method: 'GET',
                    isArray: false
                },
                GetAll: {
                    method: 'GET',
                    isArray: true
                },
                GetProject: {
                    method: 'GET',
                    isArray: true
                },
                GetProjectInactives: {
                    method: 'GET',
                    isArray: true
                },
                GetAllTestCase: {
                    method: 'GET',
                    isArray: true
                },
                GetAllTestCaseInactives: {
                    method: 'GET',
                    isArray: true
                },
                GetAllTestScenario: {
                    method: 'GET',
                    isArray: true
                },
                GetAllTestScenarioInactives: {
                    method: 'GET',
                    isArray: true
                },
                GetAllTestProcedure: {
                    method: 'GET',
                    isArray: true
                },
                GetAllTestProcedureInactives: {
                    method: 'GET',
                    isArray: true
                },
                RequirementChangeLogs: {
                    method: 'GET',
                    isArray: true
                },
                Restore: {
                    method: 'POST'
                },
                GetRequirementbyTCId: {
                    method: 'GET',
                    isArray: false
                },
                GetRequirementbyTPId: {
                    method: 'GET',
                    isArray: false
                },
                GetAllTCWithoutTP: {
                    method: 'GET',
                    isArray: true
                },
                GetAllTCWithoutTPByReq: {
                    method: 'GET',
                    isArray: true
                },
                GetEvidenceFromReq: {
                    method: 'POST'
                    , isArray: true

                },
                GetAutomatedEvidenceFromReq: {
                    method: 'POST'
                    ,isArray: true
                },
                GetRequirementsByTestEvidence: {
                    method: 'GET'
                    , isArray: true
                },
                GetSprints: {
                    method: 'GET',
                    isArray: true
                },
                GetTestCasesByNumber: {
                    method: 'POST', 
                    isArray:true
                }
            }),

            self = this;

        self.GetAll = function (success, error) {
            return resource.GetAll({ action: 'GetAll' }, success || noop, error || noop);
        };

        self.Get = function (id, success, error) {
            return resource.Get({ action: 'Get', id: id }, success || noop, error || noop);
        };

        self.Enable = function (id, success, error) {
            return resource.Enable({ action: 'Enable', id: id }, success || noop, error || noop);
        };

        self.RequirementChangeLogs = function (id, success, error) {
            return resource.RequirementChangeLogs({ action: 'RequirementChangeLogs', id: id }, success || noop, error || noop);
        };

        self.Restore = function (entity, version, success, error) {
            return resource.Restore({ action: 'Restore', version: version }, entity, success || noop, error || noop);
        };

        self.GetProject = function (Project_Id, success, error) {
            return resource.GetProject({ action: 'GetProject', Project_Id: Project_Id }, success || noop, error || noop);
        };

        self.GetSprints = function (projectId, success, error) {
            return resource.GetProject({ action: 'GetSprints', projectId: projectId }, success || noop, error || noop);
        };

        self.GetProjectInactives = function (Project_Id, success, error) {
            return resource.GetProject({ action: 'GetProjectInactives', Project_Id: Project_Id }, success || noop, error || noop);
        };

        self.GetAllTestCase = function (id, success, error) {
            return resource.GetAllTestCase({ action: 'GetAllTestCase', id: id }, success || noop, error || noop);
        };

        self.GetAllTestCaseInactives = function (id, success, error) {
            return resource.GetAllTestCaseInactives({ action: 'GetAllTestCaseInactives', id: id }, success || noop, error || noop);
        };

        self.GetAllTestScenario = function (id, success, error) {
            return resource.GetAllTestCase({ action: 'GetAllTestScenario', id: id }, success || noop, error || noop);
        };

        self.GetAllTestScenarioInactives = function (id, success, error) {
            return resource.GetAllTestScenarioInactives({ action: 'GetAllTestScenarioInactives', id: id }, success || noop, error || noop);
        };

        self.GetAllTestProcedure = function (id, success, error) {
            return resource.GetAllTestProcedure({ action: 'GetAllTestProcedure', id: id }, success || noop, error || noop);
        };

        self.GetAllTestProcedureInactives = function (id, success, error) {
            return resource.GetAllTestProcedureInactives({ action: 'GetAllTestProcedureInactives', id: id }, success || noop, error || noop);
        };

        self.Delete = function (id, success, error) {
            return resource.Delete({ action: 'Delete', id: id }, success || noop, error || noop);
        };

        self.Save = function (entity, success, error) {
            return resource.Save({ action: 'Save' }, entity, success || noop, error || noop);
        };

        self.Update = function (entity, success, error) {
            return resource.Update({ action: 'Update' }, entity, success || noop, error || noop);
        };

        self.GetRequirementbyTCId = function (id, success, error) {
            return resource.GetRequirementbyTCId({ action: 'GetRequirementbyTCId', id: id }, success || noop, error || noop);
        };

        self.GetRequirementbyTPId = function (id, success, error) {
            return resource.GetRequirementbyTPId({ action: 'GetRequirementbyTPId', id: id }, success || noop, error || noop);
        };

        self.GetAllTCWithoutTP = function (id, success, error) {
            return resource.GetAllTCWithoutTP({ action: 'GetAllTCWithoutTP', projectId: id }, success || noop, error || noop);
        };

        self.GetAllTCWithoutTPByReq = function (id, success, error) {
            return resource.GetAllTCWithoutTP({ action: 'GetAllTCWithoutTPByReq', reqId: id }, success || noop, error || noop);
        };

        self.GetEvidenceFromReq = function (id, success, error) {
            return resource.GetEvidenceFromReq({ action: 'GetEvidenceFromReq' }, id, success || noop, error || noop);
        };

        self.GetAutomatedEvidenceFromReq = function (id, success, error) {
            return resource.GetAutomatedEvidenceFromReq({ action: 'GetAutomatedEvidenceFromReq' }, id, success || noop, error || noop);
        };

        self.GetRequirementsByTestEvidence = function (id, type, success, error) {
            return resource.GetRequirementsByTestEvidence({ action: 'GetRequirementsByTestEvidence', id: id, type: type },success || noop, error || noop);
        };

        self.GetSprints = function (id, success, error) {
            return resource.GetSprints({ action: 'GetSprints', projectId: id},success || noop, error || noop);
        };

        self.GetTestCasesByNumber = function (TcNumbers, id, scriptId ,success, error) {
            return resource.GetTestCasesByNumber({ action: 'GetTestCasesByNumber', projectId: id, scriptId: scriptId}, TcNumbers,success || noop, error || noop);
        };
        
    }]);
})(angular);